diff --git a/dist/index.cjs b/dist/index.cjs
index bde0210eb131db06ccca30a5c466c7252a03af50..0f83d2a4a3237774b715c3400dc103536fff012c 100644
--- a/dist/index.cjs
+++ b/dist/index.cjs
@@ -43,8 +43,11 @@ function findBox(input, boxName, currentOffset) {
   while (currentOffset < input.length) {
     const box = readBox(input, currentOffset);
     if (!box) break;
+    if (box.size === 0) {
+      throw new TypeError("Invalid image, zero-length box");
+    }
     if (box.name === boxName) return box;
-    currentOffset += box.size > 0 ? box.size : 8;
+    currentOffset += box.size;
   }
 }
 
@@ -253,6 +256,9 @@ var ICNS = {
     const images = [];
     while (imageOffset < fileLength && imageOffset < inputLength) {
       const imageHeader = readImageHeader(input, imageOffset);
+      if (imageHeader[1] === 0) {
+        throw new TypeError("Invalid ICNS, zero-length image entry");
+      }
       const imageSize2 = getImageSize2(imageHeader[0]);
       images.push(imageSize2);
       imageOffset += imageHeader[1];
diff --git a/dist/index.mjs b/dist/index.mjs
index 22c17afad4e406eba013fff7858366cb5f8e3ba4..6f7bda1625b64770484f5ddc1581cc40a76f341b 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -39,8 +39,11 @@ function findBox(input, boxName, currentOffset) {
   while (currentOffset < input.length) {
     const box = readBox(input, currentOffset);
     if (!box) break;
+    if (box.size === 0) {
+      throw new TypeError("Invalid image, zero-length box");
+    }
     if (box.name === boxName) return box;
-    currentOffset += box.size > 0 ? box.size : 8;
+    currentOffset += box.size;
   }
 }
 
@@ -249,6 +252,9 @@ var ICNS = {
     const images = [];
     while (imageOffset < fileLength && imageOffset < inputLength) {
       const imageHeader = readImageHeader(input, imageOffset);
+      if (imageHeader[1] === 0) {
+        throw new TypeError("Invalid ICNS, zero-length image entry");
+      }
       const imageSize2 = getImageSize2(imageHeader[0]);
       images.push(imageSize2);
       imageOffset += imageHeader[1];
diff --git a/dist/types/heif.cjs b/dist/types/heif.cjs
index 1f490788ea654ea9896060a5c0253d81e348af92..11f722b669d75ad1d1ddc38194ffb206bb08614b 100644
--- a/dist/types/heif.cjs
+++ b/dist/types/heif.cjs
@@ -19,8 +19,11 @@ function findBox(input, boxName, currentOffset) {
   while (currentOffset < input.length) {
     const box = readBox(input, currentOffset);
     if (!box) break;
+    if (box.size === 0) {
+      throw new TypeError("Invalid image, zero-length box");
+    }
     if (box.name === boxName) return box;
-    currentOffset += box.size > 0 ? box.size : 8;
+    currentOffset += box.size;
   }
 }
 
diff --git a/dist/types/heif.mjs b/dist/types/heif.mjs
index 01330919e01b3c034a96a7b4b1dbf5a7f658bb2d..3f4425cc39e083778d222803c4f4f16882b1c4d1 100644
--- a/dist/types/heif.mjs
+++ b/dist/types/heif.mjs
@@ -17,8 +17,11 @@ function findBox(input, boxName, currentOffset) {
   while (currentOffset < input.length) {
     const box = readBox(input, currentOffset);
     if (!box) break;
+    if (box.size === 0) {
+      throw new TypeError("Invalid image, zero-length box");
+    }
     if (box.name === boxName) return box;
-    currentOffset += box.size > 0 ? box.size : 8;
+    currentOffset += box.size;
   }
 }
 
diff --git a/dist/types/icns.cjs b/dist/types/icns.cjs
index 32c8d370dcf80a10f14a6deebfdbe04d3c70585c..187c919d529ee0b8a42e65d53a80a8228380d687 100644
--- a/dist/types/icns.cjs
+++ b/dist/types/icns.cjs
@@ -72,6 +72,9 @@ var ICNS = {
     const images = [];
     while (imageOffset < fileLength && imageOffset < inputLength) {
       const imageHeader = readImageHeader(input, imageOffset);
+      if (imageHeader[1] === 0) {
+        throw new TypeError("Invalid ICNS, zero-length image entry");
+      }
       const imageSize = getImageSize(imageHeader[0]);
       images.push(imageSize);
       imageOffset += imageHeader[1];
diff --git a/dist/types/icns.mjs b/dist/types/icns.mjs
index 8710027bc5890a79d05a7fdb87de3018e556729e..ecf8ffe7c1e3f28b01af8c5b097b308a45bf92ea 100644
--- a/dist/types/icns.mjs
+++ b/dist/types/icns.mjs
@@ -70,6 +70,9 @@ var ICNS = {
     const images = [];
     while (imageOffset < fileLength && imageOffset < inputLength) {
       const imageHeader = readImageHeader(input, imageOffset);
+      if (imageHeader[1] === 0) {
+        throw new TypeError("Invalid ICNS, zero-length image entry");
+      }
       const imageSize = getImageSize(imageHeader[0]);
       images.push(imageSize);
       imageOffset += imageHeader[1];
diff --git a/dist/types/jxl.cjs b/dist/types/jxl.cjs
index a12095be85c7a8fd25763895a47c46b4292ebd9c..4fd569c5a6a0685476930946e3f46e54081179df 100644
--- a/dist/types/jxl.cjs
+++ b/dist/types/jxl.cjs
@@ -60,8 +60,11 @@ function findBox(input, boxName, currentOffset) {
   while (currentOffset < input.length) {
     const box = readBox(input, currentOffset);
     if (!box) break;
+    if (box.size === 0) {
+      throw new TypeError("Invalid image, zero-length box");
+    }
     if (box.name === boxName) return box;
-    currentOffset += box.size > 0 ? box.size : 8;
+    currentOffset += box.size;
   }
 }
 
diff --git a/dist/types/jxl.mjs b/dist/types/jxl.mjs
index 49b0c4cac6abb366378e5feed745f16981b441c0..c31e78e2cf5f804f12f606f15a8421ad7b5c1dca 100644
--- a/dist/types/jxl.mjs
+++ b/dist/types/jxl.mjs
@@ -58,8 +58,11 @@ function findBox(input, boxName, currentOffset) {
   while (currentOffset < input.length) {
     const box = readBox(input, currentOffset);
     if (!box) break;
+    if (box.size === 0) {
+      throw new TypeError("Invalid image, zero-length box");
+    }
     if (box.name === boxName) return box;
-    currentOffset += box.size > 0 ? box.size : 8;
+    currentOffset += box.size;
   }
 }
 
